### PR TITLE
SQL-2094: Run static code analysis for ODBC

### PR DESCRIPTION
**Pull request by EthanHardyMongo**
Successful patch: https://spruce.mongodb.com/task/mongosql_odbc_driver_code_quality_security_semgrep_patch_348c654cbebe856370bf8415feaa0edd0b20c7f8_66620494fb55da00077021a2_24_06_06_18_48_53/logs?execution=0

verbose output: https://mciuploads.s3.amazonaws.com/mongosql-odbc-driver/artifacts/66620494fb55da00077021a2/code-quality-security/ssdlc/mongo-odbc-driver.sast.cmd.verbose.out

sarif file: https://mciuploads.s3.amazonaws.com/mongosql-odbc-driver/artifacts/66620494fb55da00077021a2/code-quality-security/ssdlc/mongo-odbc-driver.sast.sarif

### CODE REVIEW

**Comment by EthanHardyMongo** [comment]
Everything seems to be working except that I am getting 0 code rules.

```
  Scanning 165 files tracked by git with 0 Code rules:
  Nothing to scan.
If Semgrep missed a finding, please send us feedback to let us know!
See https://semgrep.dev/docs/reporting-false-negatives/
Sending pseudonymous metrics since metrics are configured to AUTO and registry usage is True
{
  "$schema": "https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/schemas/sarif-schema-2.1.0.json",
  "runs": [
    {
      "invocations": [
        {
          "executionSuccessful": true,
          "toolExecutionNotifications": []
        }
      ],
      "results": [],
      "tool": {
        "driver": {
          "name": "Semgrep OSS",
          "rules": [],
          "semanticVersion": "1.75.0"
        }
      }
    }
  ],
  "version": "2.1.0"
}
```

File: evergreen.yml [hunk 1]
```diff
@@ -1582,6 +1582,47 @@ functions:
...
+    - command: shell.exec
+      type: test
+      params:
+        shell: bash
+        working_dir: mongosql-odbc-driver
+        script: |
+          ${prepare_shell}
+          export SEMGREP_APP_TOKEN=${semgrep_app_token}
+          echo "Running static code analysis with Semgrep..."
+          venv='venv'
```
Comment (mattChiaravalloti) [inline evergreen.yml:1595]: What's the benefit of this?

File: evergreen.yml [hunk 2]
```diff
@@ -1582,6 +1582,47 @@ functions:
...
+              echo 'using existing virtualenv'
+              . "$venv"/bin/activate
+          else
+              echo 'Creating new virtualenv'
+              python3 -m virtualenv "$venv"
+              echo 'Activating new virtualenv'
+              . "$venv"/bin/activate
+          fi
+          python3 -m pip install semgrep
+          # confirm
```
Comment (mattChiaravalloti) [inline evergreen.yml:1607]:
```suggestion
          # Confirm semgrep version
```
Let's expand this comment to include more meaning.

**Review by mattChiaravalloti — CHANGES_REQUESTED** [review]

File: evergreen.yml (hunk 1 above)
Comment (nbagnard) [reply evergreen.yml:1595]: ahhh, if you are really interested you can read at this thread: https://mongodb.slack.com/archives/C05N744KWQ4/p1717592697786179

Short explanation if you don't want to get bogged down with details : Semgrep has 2 different set of rules. Community rules and Pro rules. For Pro rules to be used, you need a Semgrep account and generate an API token.

For Rust, useful rules are only Pro rules. So, without an account, we had no rules applied to the scan (which yield a perfectly clean result of no vulnerabilities :P )

Comment (mattChiaravalloti) [reply evergreen.yml:1595]: I was specifically talking about `venv='venv'` not the `SEMGREP_APP_TOKEN`

Comment (EthanHardyMongo) [reply evergreen.yml:1595]: When I was making this script, I was largely just copying what had already be done in the BI-Connector and JDBC. This was done in both those repos (These PRs show it: https://github.com/10gen/sqlproxy/pull/453 and https://github.com/mongodb/mongo-jdbc-driver/pull/261), so I included it here, but I'm not actually sure what the point of it is. It doesn't seem necessary.

Comment (EthanHardyMongo) [reply evergreen.yml:1595]: I ran a patch without it, it seems to work fine: https://spruce.mongodb.com/version/666202b446b7c80007c8f008/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

new code:
```
          echo "Running static code analysis with Semgrep..."

          # Setup or use the existing virtualenv for semgrep
          if [[ -f "venv/bin/activate" ]]; then
              echo 'using existing virtualenv'
              . venv/bin/activate
          else
              echo 'Creating new virtualenv'
              python3 -m virtualenv venv
              echo 'Activating new virtualenv'
              . venv/bin/activate
          fi
          python3 -m pip install semgrep
```

**Review by mattChiaravalloti — APPROVED** [review]

File: evergreen.yml [hunk 3]
```diff
@@ -1582,6 +1582,47 @@ functions:
...
+              echo 'Creating new virtualenv'
+              python3 -m virtualenv venv
+              echo 'Activating new virtualenv'
+              . venv/bin/activate
+          fi
+          python3 -m pip install semgrep
+          # Confirm semgrep version
+          semgrep --version
+          set +e
+          semgrep --config p/rust --sarif --exclude "integration_test" --verbose --error --severity=ERROR --sarif-output=mongo-odbc-driver.sast.sarif > mongo-odbc-driver.sast.cmd.verbose.out 2>&1
```
Comment (nbagnard) [inline evergreen.yml:1610]: I looked at the SSDLC policy and Semgrep documentation again and we should be running all rules, not only the ones with severity ERROR. I'll make follow up tickets. We need to ignore rules like "Should use rust unsafe code" for example and find a way to tag other findings with False Positive or No fix needed.

**Review by nbagnard — APPROVED** [review]
LGTM! 👍

**Pull request merged**